Redesigned UsbCover (and more)

gluap opened the pull request:
I find the work done by @opatut on the OpenSCAD-Case impressive, and I love that we finally have an easy accessible open source based design for the Case. 

This is a different design of the UsbCover - Fits the current version of the case.

As discussed. When I started it I didn't know @opatut was already working on the same thing as it was still marked as "todo" in the forum. Should've announced I'm working on it right when I started.

Clearance is generated via Minkowski with a cylinder.

Further changes to other parts:
- add clearance to fit sensor board
- Chamfer for lid edge.

![image](https://user-images.githubusercontent.com/44007906/148250175-60216844-476a-4c16-a11a-2a0ff05f1d9f.png)
![works](https://user-images.githubusercontent.com/44007906/148250192-64fff2df-43ca-4572-a3eb-db6d9c8adb16.jpeg)

gluap commented on src/MainCase/MainCase.scad:
```diff
@@ -266,8 +266,8 @@ module MainCase(without_inserts=false) {
               [UsbCharger_width, 0, 0],
-              [UsbCharger_width, UsbCharger_depth, 3],
-              [0, UsbCharger_depth, 0],
+              [UsbCharger_width, UsbCharger_depth+wall_thickness/2, 3],
```
Otherwise the wall is not even 0.4mm wide and not printed.

opatut replied:
The wall there needs to be very thin, as it will prevent the USB-C plug from going in deep enough if it is too thick. Please make it a constant 0.8mm or so (instead of depending on wall_thickness). That will be 2 traces at 0.4mm nozzle, or one trace if printing with a wider nozzle. I don't think we can afford much more.

gluap commented on src/MainCase/MainCase.scad:
```diff
@@ -236,9 +236,9 @@ module MainCase(without_inserts=false) {
           polygon(polyRound([
             [wall_thickness, wall_thickness+HeatsetInsert_height, 0],
             [OBS_height-wall_thickness, wall_thickness+HeatsetInsert_height, 0],
-            [OBS_height-wall_thickness, OBS_depth-20, 0],
```
Clearance in this corner was required for the sensor board
![image](https://user-images.githubusercontent.com/44007906/148275155-96bd5ba7-f886-4da5-b6bb-d626e01a38eb.png)

opatut replied:
The heatsert inserts from the top might stick through there. Don't assume MainCase to be complete and working as of yet ;)

gluap commented on src/MainCase/MainCaseLid.scad:
```diff
@@ -58,20 +59,20 @@ module MainCaseLid() {
       translate([36, 39, 0])
       rotate([0, 0, 90])
       BatteryHolder();
+      minkowski(){
```
minkowski allows for an easy chamfer without causing too much complexity when using a pyramid. 
![image](https://user-images.githubusercontent.com/44007906/148275302-abe6be0c-ea23-43c1-a5d1-427385f90045.png)

opatut replied:
Yeah, but it is super slow :( I had a minkowski there earlier and removed it because I was unhappy with it.

gluap replied:
A test run without this minkowski and flushed caches for me takes 24 seconds to build from scratch.
A test run with this minkowski and flushed caches takes about 55 seconds. 

I suggest we handle this by only using time-intensive the chamfer pyramid in non-preview mode - during stl build I don't think the compute time is too much to ask. I'll add that change.

I also explored using the trick with a shifted, smaller polygon and hull, but because this shape has a hole in it that trick isn't as easily applied. It would be possible by designing an (in the mathematical sense) star-shaped cutter to cut the hole, but Imho the toll taken in readability would not justify the gains of a factor two in build time.

opatut commented on lib/utils.scad:
```diff
@@ -46,3 +46,11 @@ function roundedRectangle(width, height, radius) =
   ], fn=$pfn);
+
+
+module ChamferPyramid(l)       polyhedron(
+  points=[ [l,l,0],[l,-l,0],[-l,-l,0],[-l,l,0], 
+           [0,0,l]  ],                                 
+  faces=[ [0,1,4],[1,2,4],[2,3,4],[3,0,4],             
+              [1,0,3],[2,1,3] ]                     
+ );
```
Could you format this more nicely?

opatut commented on src/MainCase/USBCover.scad:
```diff
@@ -0,0 +1,119 @@
+include <../../variables.scad>
```
Duplicate file?

opatut commented on variables.scad:
```diff
@@ -14,23 +14,26 @@ SensorHole_ledge = 1.0;
+UsbCover_r1 = 1.8; // the rounded corners with smaller radius
+UsbCover_r2 = 2; // the rounded corners with bigger radius
+UsbCover_magnet_size = 3.0; // decrease if magnets are too loose
+UsbCover_magnet_clearance = 0.25; // increase if your printer needs more clearance for press fit
```
I don't understand why the clearance and the size are modelled differently. Why not just increase the magnet size if your hole is too narrow on your printer?

gluap replied:
The trick is that it is much easier to have a narrow opening but a wide hole for the magnets: you just have to squeeze the magnets through the narrow hole at the top after which they comfortably sit in their roomy hole with plenty clearance. This way the opening can be much tighter and will grip the magnet. More detailed discussion [here](https://github.com/openbikesensor/OpenBikeSensor3dPrintableCase/issues/14)

opatut commented on variables.scad:
```diff
@@ -14,23 +14,26 @@ SensorHole_ledge = 1.0;
 UsbCover_indent_depth = 1.5;
-UsbCover_indent_radius = 0.85;
 UsbCover_indent_height = 4;
 UsbCover_indent_width = 12;
 
-UsbCover_wing_offset = 1.88;
-UsbCover_wing_width = 2.75;
+UsbCover_wing_offset = 2.2;
+UsbCover_wing_width = 2.6;
```
Why are these values so different from what I measured?

I might need to testprint this.

gluap replied:
Because we were looking at two different initial designs - I was using the latest version and had to measure on a ``.stl`` which is cumbersome, you went around that part by using the ``.step`` file from an "older" folder.

The easiest way to verify this matches what we currently have is to subtract this model and the latest .stl in openscad - I will add a line to that end.

gluap replied:
![image](https://user-images.githubusercontent.com/44007906/148612800-c5ca7614-cb49-4920-8d72-88ef7167a5b5.png)
![image](https://user-images.githubusercontent.com/44007906/148612883-bfa16504-14c6-4bae-96c8-2a7f0b1a21d1.png)

compared using the following import lines in UsbCover.scad
```
translate([UsbCover_width+5,0,0])translate([-50.1,UsbCover_depth,0])rotate([0,270,90])import("../../MainCase/VerticalCase_JSN-AJ/OBS-MainCase-A-003_USB-Cover_v0.1.1.stl");
translate([-UsbCover_width-5,0,0])translate([-UsbCover_width/2,0.15,0])import("../../MainCase/VerticalCase_JSN-AJ/OldVersions/OBS-MainCase-A-003_USB-Cover_v0.1.0.stl");
```

gluap commented:
@opatut I took a shot at addressing the points you made in your review. Thank you for the thorough review, your feedback was to the point as usual and is very appreciated.

opatut commented on src/MainCase/MainCase.scad:
```diff
@@ -3,7 +3,7 @@ include <../../variables.scad>
 use <../../lib/MountRails.scad>
 use <../../lib/utils.scad>
-use <./UsbCover.scad>
+use <./UsbCover.scad> 
```
Please configure your editor to remove empty whitespace at EOL.

gluap replied:
I was using OpenSCAD so far for obvious reasons and didn't see that option. Removed this whitespace from the PR though.

opatut replied:
You can use OpenSCAD in "Automatic reload and preview" mode and then use your normal editor for writing the code ;)

opatut commented on src/MainCase/MainCase.scad:
```diff
@@ -266,8 +266,8 @@ module MainCase(without_inserts=false) {
-              [UsbCharger_width, UsbCharger_depth+wall_thickness/2, 3],
-              [0, UsbCharger_depth+wall_thickness/2, 0],
+              [UsbCharger_width, UsbCharger_depth+2*extrude_width, 3],
+              [0, UsbCharger_depth+2*extrude_width, 0],
```
I don't know -- this is dangerous, if somebody has a 0.8 nozzle (@tobst does this, I think) then this will be 1.6mm, which is way too much I fear. We should try to pick the smallest multiple of `extude_width` that is `>= 0.6` or so. This might work:

```openscad
ceil(0.6 / extrude_width) * extrude_width
```

gluap replied:
it doesn't have to be a multiple necessarily, but with .4 mm nozzles a single-extrusion wall is dangerous.

we can also just leave it at .92 mm fixed - any extrusion width > .46 will then produce a single wall extrusion.
